qiskit/pulse/transforms/resolved_frame.py

```diff
@@ -0,0 +1,210 @@
+# that they have been altered from the originals.
+
+"""Implements a Frame."""
+
+from abc import ABC
+from typing import Optional
+import numpy as np
+
+from qiskit.circuit.parameterexpression import ParameterExpression
+from qiskit.pulse.channels import PulseChannel
+from qiskit.pulse.frame import Frame
+from qiskit.pulse.schedule import Schedule
+from qiskit.pulse.instructions.frequency import SetFrequency, ShiftFrequency
+from qiskit.pulse.instructions.phase import SetPhase, ShiftPhase
+from qiskit.pulse.exceptions import PulseError
+
+
+class Tracker(ABC):
+    """
+    Implements a class to keep track of the phase and frequency of a frame
+    or a pulse channel in a given schedule.
+    """
+
+    def __init__(self, identifier: str, sample_duration: float):
+        """
+        Args:
+            identifier: The identifier of the Tracker.
+            sample_duration: Duration of a sample.
+        """
+        self._identifier = identifier
+        self._frequencies_phases = []  # List of (time, frequency, phase) tuples
+        self._instructions = {}
+        self._sample_duration = sample_duration
+
+    @property
+    def identifier(self) -> str:
+        """Return the identifier of the tracker."""
+        return self._identifier
+
+    def frequency(self, time: int) -> float:
+        """
+        Get the most recent frequency of the tracker before time.
+
+        Args:
+            time: The maximum time for which to get the frequency.
+
+        Returns:
+            frequency: The frequency of the frame right before time.
+        """
+        frequency = self._frequencies_phases[0][1]
+        for time_freq in self._frequencies_phases:
+            if time_freq[0] <= time:
+                frequency = time_freq[1]
+            else:
+                break
+
+        return frequency
+
+    def phase(self, time: int) -> float:
+        """
+        Get the most recent phase of the tracker.
+
+        Args:
+            time: The maximum time for which to get the phase.
+
+        Returns:
+             phase: The phase of the frame up until time.
+        """
+        if len(self._frequencies_phases) == 0:
+            return 0.0
+
+        phase = self._frequencies_phases[0][1]
+        last_time = self._frequencies_phases[0][0]
+
+        for time_freq in self._frequencies_phases:
+            if time_freq[0] <= time:
+                phase = time_freq[2]
+                last_time = time_freq[0]
+            else:
+                break
+
+        freq = self.frequency(time)
+
+        return (phase + 2 * np.pi * freq * (time - last_time) * self._sample_duration) % (2 * np.pi)
+
+    def set_frequency(self, time: int, frequency: float):
+        """Insert a new frequency in the time-ordered frequencies.
+
+        Args:
+            time: The time in samples (i.e. measured in units of dt).
+            frequency: The frequency to which self is set after the given time.
+        """
+        insert_idx = 0
```

Comment:
I think you can just do below, rather than manually looping over all entries.
```
self._frequencies_phases = sorted(self._frequencies_phases + [new_tuple], key=lambda x: x[0])
```

I guess `sorted` is implemented by C so much faster than python loop.

Reply:
+1

Reply:
Sounds good. See here: 31ef605